### ISSUE TITLE
Add assert() function

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#301]
Assertions are a useful debugging tool.  Perhaps add an `assert()` function to Duktape which pauses execution on failure if the debugger is attached and is a no-op otherwise.

Downside: Unlike a C ASSERT macro, the expression is always evaluated.

**Comment by svaarala** [comment]
It would be possible to write an assert function which was "understood" by the compiler so that compiler could omit the call entirely at compile time, or emit a special opcode which allowed the compiler to skip the expression if assertions were not enabled at run time.

This ties in a bit with optimizing external function access when (1) the scope path can be analyzed at compile time, and (2) the binding where the function is found is non-writable and non-configurable. In such cases we don't need a slow run-time reference but could directly reference the function.

If the compiler had that capability, it could also check if the function is "special". For example, if a function had the internal property `_Disable` (or whatever the name was), the compiler could omit the expression and not call the function; the call expression result could always be `undefined` if that matters. Rather than the `assert()` function being magic, that capability would be available for arbitrary user debug functions too.

I've worked with Lua in a similar situation and the solution we used for this kind of thing was bytecode compilation with a modified compiler which was tweaked to drop function calls we didn't want for release builds. The same would work for Duktape, but it'd be nice if there'd be a cleaner solution than this.

**Comment by fatcerberus** [comment]
So something like intrinsic functions in a C/C++ compiler, where the compiler recognizes certain function calls (such as memcpy) and substitutes them with direct, inline variants?

**Comment by svaarala** [comment]
Not quite: they would be called normally, but the compiler would just have the necessary information to decide, at compile time, that the calls are intended to be omitted for that function.

**Comment by svaarala** [comment]
Actually "intrinsics" would also possible to add: one would register identifiers that would be recognized by the compiler in any function with a direct call to the function registered with it. The compiler would check the intrinsics in identifier resolution handling.

But I'd probably prefer the compiler do the actual identifier resolution at compile time and *when* it was safe to assume that target function will always be the same at run time, only then optimize for it. That way the literal source doesn't have hidden assumptions about certain bindings being magically available.

**Comment by fatcerberus** [comment]
Yeah, that definitely makes sense.  Developers expect asserts to be magic, everything else is expected to be at least somewhat deterministic. :)

**Comment by fatcerberus** [comment]
Another related enhancement would be a "debug print" function which only produced output in the debug client (not the console as print() does).

**Comment by svaarala** [comment]
Second that :)

**Comment by fatcerberus** [comment]
There is an issue with having the compiler specifically recognize calls to `assert()` and similar, and that's that any code which uses such names for its own functions is likely to break.  This would require a bit of detection so that, if these names are reassigned, the compiler would stop treating them specially.

**Comment by svaarala** [comment]
Yeah, that's what I suggested with the compiler doing the identifier lookup at compile time to figure out which binding it will necessarily hit at run time. It cannot always be determined at compile time though.

**Comment by fatcerberus** [comment]
> It cannot always be determined at compile time though.

Once again `eval` is the bane of everyone's existence. ;)

**Comment by svaarala** [comment]
There is also the "with" binding which can capture the access.

**Comment by svaarala** [comment]
This is how:

```js
function assert() {}

with ({ assert: 123 }) {
    print(typeof assert);  // prints 'number'
}
```

And the with target can naturally be an arbitrary object coming from outside.

**Comment by fatcerberus** [comment]
Yeah, the JS implementation of `with` always baffled me.  It's dumb in that it redefines the semantics of everything inside of it, creating ambiguity and making the compiler's job more difficult.

This would have been much better:
```javascript
with (pig) {
  .sitOn("Scott");
  .eat("Scott");
}
```

Alas, it's far too late for that now. :(

**Comment by fatcerberus** [comment]
Hm, now that I think of it, would it be possible to get the "debug print" functionality (print only to debugger, not stdout) using `Duktape.Logger`?

**Comment by markand** [comment]
While I would like to see an assert function because it requires some intrinsics internals (line number and expression content), I would like to get it in Duktape object and not as global function since it is not present in Ecmascript specifications :)

**Comment by fatcerberus** [comment]
I was thinking the same thing, however I'd be curious if there's any precedent for a top-level ‘assert()` function - `print()` and `alert()` aren't part of the Ecmascript spec either but Duktape provides them anyway because they are present in a browser environment. ;)

**Comment by svaarala** [comment]
@fatcerberus Those two are, I think, also worth reconsidering whether they should be present by default. The current situation is not entirely consistent because e.g. `console` is often expected but not provided by default. The division between Duktape vs. global, on the other hand, is mostly based on whether the function has only relevance to Duktape (but even there `Duktape.Logger` is more or less generic functionality rather than related to e.g. finalizers or types).

So all in all, we can change things, hopefully for the better, in 2.x or later major versions.

**Comment by svaarala** [comment]
One case for including `print()` is that it makes it easy to write tests and command line tests. But then, it would also be quite useful for have minimal `readFile()` and `writeFile()` for that, which are already a bit more problematic.

So my instinct would be to remove `print()` and `alert()` from the core built-ins because they have I/O portability issues and don't need anything "magic" so they can be removed. We can provide them as a separate file in the distributable of course - perhaps with a `console` binding also included.
